File: capstone-projects/cafe-management-program/cpp/Project_Nhom04_Cpp/Date.cpp
#include "Library.h"

//properties
int Date::getiDay() {
	return iDay;
}

void Date::setiDay(int iDay) {
	this->iDay = iDay;
}

int Date::getiMonth() {
	return iMonth;
}

void Date::setiMonth(int iMonth) {
	this->iMonth = iMonth;
}

int Date::getiYear() {
	return iYear;
}

void Date::setiYear(int iYear) {
	this->iYear = iYear;
}

Date::Date() {
}

Date::~Date() {
}

void Date::Input() {
	string a;
	Date b;

	while (true) {
		try {

			cout << "Date [dd/mm/yy]: ";
			getline(cin, a);

			b = ToDate(a);

			if (!IsDate(b.iDay, b.iMonth, b.iYear))
				continue;
			else
			{
				this->iDay = b.iDay;
				this->iMonth = b.iMonth;
				this->iYear = b.iYear;
				break;
			}
		}
		catch (exception e) {
			continue;
		}
	}
}

//Output
void Date::Output() {
	cout << to_string(this->iDay) + "/" + to_string(this->iMonth) + "/" + to_string(this->iYear);
}

//DATE methods
//Operators
bool IsEqual(Date a, Date b) {
	return (a.getiDay() == b.getiDay()) && (a.getiMonth() == b.getiMonth()) && (a.getiYear() == b.getiYear());
}

bool IsMore(Date a, Date b) {
	if (a.getiYear() > b.getiYear()) return true;
	else if (a.getiYear() < b.getiYear()) return false;

	if (a.getiMonth() > b.getiMonth()) return true;
	else if (a.getiMonth() < b.getiMonth()) return false;

	return a.getiDay() > b.getiDay();
}

string ToString(Date a) {
	string d = "";
	d += to_string(a.getiDay()) + "/" + to_string(a.getiMonth()) + "/" + to_string(a.getiYear());
	return d;
}

bool IsLeapYear(int y) {
	return (y % 4 == 0) && (y % 100 != 0 || y % 400 == 0);
}

int DaysInMonth(int m, int y) {
	if (m == 4 || m == 6 || m == 9 || m == 11) return 30;
	else if (m == 2) {
		if (IsLeapYear(y)) return 29;
		else return 28;
	}
	else return 31;
}

bool IsDate(int Day, int Month, int Year) {
	return Day > 0 && Day <= DaysInMonth(Month, Year) && Month > 0 && Month < 13 && Year > 0;
}

Date ToDate(string a) {
	Date d;
	string split[MAX];
	SplitString(split, a, '/');
	d.setiDay(stoi(split[0]));
	d.setiMonth(stoi(split[1]));
	d.setiYear(stoi(split[2]));
	return d;
}

File: capstone-projects/cafe-management-program/java/src/Date.java
import java.util.*;

public class Date {
    //Fields
    private int iDay;
    private int iMonth;
    private int iYear;

    public int getiDay() {
        return iDay;
    }

    public void setiDay(int iDay) {
        this.iDay = iDay;
    }

    public int getiMonth() {
        return iMonth;
    }

    public void setiMonth(int iMonth) {
        this.iMonth = iMonth;
    }

    public int getiYear() {
        return iYear;
    }

    public void setiYear(int iYear) {
        this.iYear = iYear;
    }

    //Constructors
    public Date() {
    }

    //Input
    public void Input() {
        Scanner sc = new Scanner(System.in);
        String a;
        String[] b;
        do {
            try {
                System.out.print("Date [dd/mm/yyyy]: ");
                a = sc.nextLine();
                b = a.split("/");

                this.iDay = Integer.parseInt(b[0]);
                this.iMonth = Integer.parseInt(b[1]);
                this.iYear = Integer.parseInt(b[2]);
                if (!IsDate(this.iDay, this.iMonth, this.iYear)) continue;
                else break;
            } catch (Exception e) {
                continue;
            }
        } while (true);
    }

    //Output
    public void Output() {
        System.out.println(this.iDay + "/" + this.iMonth + "/" + this.iYear);
    }

    //DATE methods
    static public boolean IsDate(int Day, int Month, int Year) {
        return Day > 0 && Day <= DaysInMonth(Month, Year) && Month > 0 && Month < 13 && Year > 0;
    }

    static boolean IsLeapYear(int y) {
        return (y % 4 == 0) && (y % 100 != 0 || y % 400 == 0);
    }

    static int DaysInMonth(int m, int y) {
        if (m == 4 || m == 6 || m == 9 || m == 11) return 30;
        else if (m == 2) {
            if (IsLeapYear(y)) return 29;
            else return 28;
        } else return 31;
    }

    //Operators
    public static boolean IsEqual(Date a, Date b) {
        return (a.getiDay() == b.getiDay()) && (a.getiMonth() == b.getiMonth()) && (a.getiYear() == b.getiYear());
    }

    public static boolean IsMore(Date a, Date b) {
        if (a.getiYear() > b.getiYear()) return true;
        else if (a.getiYear() < b.getiYear()) return false;

        if (a.getiMonth() > b.getiMonth()) return true;
        else if (a.getiMonth() < b.getiMonth()) return false;

        return a.getiDay() > b.getiDay();
    }

    public static String ToString(Date a) {
        String day = "";
        day += a.getiDay() + "/" + a.getiMonth() + "/" + a.getiYear();
        return day;
    }

    public static Date ToDate(String a) {
        Date d = new Date();
        String[] b = a.split("/");

        d.setiDay(Integer.parseInt(b[0]));
        d.setiMonth(Integer.parseInt(b[1]));
        d.setiYear(Integer.parseInt(b[2]));

        return d;
    }
}
